php/libraries/NDB_Factory.class.inc

```diff
@@ -194,17 +193,19 @@ class NDB_Factory
             if ($db_ref !== null) {
                 return $db_ref;
             }
-                self::$db = Database::singleton();
+            self::$db = Database::singleton();
         }
         $config = $this->config();
         $dbc    = $config->getSetting('database');
-        $db_ref->connect(
-            $dbc['database'],
-            $dbc['username'],
-            $dbc['password'],
-            $dbc['host'],
-            true
-        );
+        if (!is_null($db_ref)) {
+            $db_ref->connect(
+                $dbc['database'],
+                $dbc['username'],
+                $dbc['password'],
+                $dbc['host'],
+                true
+            );
+        }
```

Comment:
I don't understand this. Given the function signature, I only see 2 possibilities:

1. $db_ref is never null (and determined as such by static analysis), in which case the if statement is meaningless.
2. $db_ref can be null in some code paths (and is determined as such by static analysis), in which case the return line below should get flagged as being incompatible with the function signature (which isn't nullable)

Reply:
I don't understand it either but it keeps coming up. It's somehow an issue existing in the code already but doesn't get flagged by `phan`.

A similar thing happened in my other PR here:
https://github.com/aces/Loris/pull/4914/files#diff-f89f7046f643b2862edaf5cce173dc17R111

Phan always thinks this is null.

I think it have some sort of troubles working with references that are set to `null` initially and later kind of juggled around in this way.

I tried updating phan to the latest version locally but the issue still comes up so it's not a bug (or at least not one they're aware of or disagree with).

I'm open to suggestions, either a different refactoring approach (like in the PR I linked) or to adding a suppression here.

Let me know what you want to do.

Reply:
Would phan be happy if you refactored it to

```
if ($db_ref === null) {
     throw Exception
}
$db_ref->connect();
return $db_ref;
```
?

Reply:
I think so. I'll change it.